### One-Way ANOVA

In the following sections, we will use clustering to analyze how the clusters differ in demographic data.

$y_{ij} = μ + \beta_{i} + ϵ_{ij}$

where:
- $i$ denotes the i-th cluster,
- j denotes the j-th tract,
- $\beta_i$ represents the effect of the i-th cluster (i.e., the difference between clusters), and
- $ϵ_{ij}$ represents the residuals within clusters.



In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
import statsmodels.formula.api as smf
import requests

import scipy.stats as stats
csvUrl = 'https://raw.githubusercontent.com/victorDeveloper0821/Projects-Spring-2025/refs/heads/main/Displaced%20Voices/Data/df_eviction_full.csv'
fullDataset = pd.read_csv(csvUrl)


fullDataset = fullDataset.drop(columns=['medianIncome_scaled','filedEviction_scaled'])

fullDataset.head()


,GEOID,zip_code,filedEviction,medianIncome,state,county,tract,incomeLevel,Single-family homes,Small multi-unit buildings (2-4 units),...,White Alone,Black Alone,Hispanic or Latino,American Indian or Alaska Native Alone,Asian Alone,Native Hawaiian or Pacific Islander Alone,Some other race,"Mixed race (exluding ""Some other race"")",Majority,cluster
0,4013010102,85331.0,1,188486,4,13,10102,Very High,0.967068,0.0,...,0.916840,0.005427,0.085874,0.0,0.012291,0.0,0.002235,0.020910,White Alone,8
1,4013010103,85262.0,1,117813,4,13,10103,Very High,0.890501,0.0,...,0.918229,0.003803,0.076881,0.0,0.000000,0.0,0.004890,0.057865,White Alone,0
2,4013010104,NaN,0,140587,4,13,10104,Very High,0.967761,0.0,...,0.955605,0.000000,0.010220,0.0,0.007027,0.0,0.000000,0.025232,White Alone,0
3,4013030401,85331.0,1,145865,4,13,30401,Very High,0.904664,0.0,...,0.912521,0.005691,0.017707,0.0,0.025295,0.0,0.000000,0.043634,White Alone,0
4,4013030401,85377.0,1,145865,4,13,30401,Very High,0.904664,0.0,...,0.912521,0.005691,0.017707,0.0,0.025295,0.0,0.000000,0.043634,White Alone,0


# **medianIncome**

In [17]:
median_anova = smf.ols(formula='filedEviction ~ C(cluster) + medianIncome * C(incomeLevel)', data=fullDataset).fit()
anova_table = sm.stats.anova_lm(median_anova, typ=2)
print(anova_table)

                                   sum_sq      df            F    PR(>F)
C(cluster)                   4.738730e+06     8.0  1930.307903  0.000000
C(incomeLevel)               3.150609e+03     3.0     3.422378  0.016664
medianIncome                 1.581198e+02     1.0     0.515277  0.472974
medianIncome:C(incomeLevel)  5.081887e+03     3.0     5.520246  0.000905
Residual                     4.655120e+05  1517.0          NaN       NaN


**Description**:
In the anova, filed eviction cases are statistically affected by cluster and interaction of `medianIncome` and `incomeLevel`.

# LiveType

In [18]:
eviction_anova = smf.ols(formula='filedEviction ~ C(cluster) + OwnerOccupied + RenterOccupied + C(LiveType):C(cluster)', data=fullDataset).fit()
anova_table = sm.stats.anova_lm(eviction_anova, typ=2)
print(anova_table)
print(eviction_anova.summary())

                              sum_sq      df            F    PR(>F)
C(cluster)              2.655280e+06     8.0  1083.008677  0.000000
C(LiveType):C(cluster)  1.273128e+04     9.0     4.615736  0.000014
OwnerOccupied           1.130279e+03     1.0     3.688055  0.054991
RenterOccupied          5.172403e+02     1.0     1.687733  0.194098
Residual                4.643026e+05  1515.0          NaN       NaN
                            OLS Regression Results                            
Dep. Variable:          filedEviction   R-squared:                       0.918
Model:                            OLS   Adj. R-squared:                  0.918
Method:                 Least Squares   F-statistic:                     1004.
Date:                Wed, 30 Apr 2025   Prob (F-statistic):               0.00
Time:                        17:20:07   Log-Likelihood:                -6554.5
No. Observations:                1533   AIC:                         1.314e+04
Df Residuals:                    1515  

/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 7
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 9, but rank is 8
  warnings.warn('covariance of constraints does not have full '


**Description**: In the anova, we can see ratios of owners and renters between clusters are not strong evidences to filed eviction cases in each cluster.

# Burden

In [19]:
burden_anova = smf.ols(formula='filedEviction ~ C(cluster) + lowBurdan + moderateBurdan + costBurdan + highBurdan + C(burden)+C(cluster):C(burden)', data=fullDataset).fit()
anova_table = sm.stats.anova_lm(burden_anova, typ=2)
print(anova_table)

/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 4
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 24, but rank is 20
  warnings.warn('covariance of constraints does not have full '


                            sum_sq      df            F         PR(>F)
C(cluster)            3.206225e+06     8.0  1285.763584   0.000000e+00
C(burden)             4.384829e+05     3.0   468.908878  9.201499e-215
C(cluster):C(burden)  7.627144e+03    24.0     1.019548   4.347135e-01
lowBurdan             1.284593e+03     1.0     4.121190   4.252580e-02
moderateBurdan        2.076910e+03     1.0     6.663077   9.937462e-03
costBurdan            1.170178e+03     1.0     3.754128   5.286514e-02
highBurdan            1.009461e+03     1.0     3.238519   7.212691e-02
Residual              4.666215e+05  1497.0          NaN            NaN


**Description:**The anova results showed that filed eviction cases are affected by compositions of rent burden with moderateBurden, costBurden and lowBurden.

# Majority

In [20]:
fullDataset.columns

Index(['GEOID', 'zip_code', 'filedEviction', 'medianIncome', 'state', 'county',
       'tract', 'incomeLevel', 'Single-family homes',
       'Small multi-unit buildings (2-4 units)',
       'Larger apartment complexes (5+ units)',
       'Mobile homes, boats, RVs, etc.', 'majorHouseType', 'lowBurdan',
       'moderateBurdan', 'costBurdan', 'highBurdan', 'burden', 'OwnerOccupied',
       'RenterOccupied', 'LiveType', 'White Alone', 'Black Alone',
       'Hispanic or Latino', 'American Indian or Alaska Native Alone',
       'Asian Alone', 'Native Hawaiian or Pacific Islander Alone',
       'Some other race', 'Mixed race (exluding "Some other race")',
       'Majority', 'cluster'],
      dtype='object')

In [21]:
ethical = fullDataset[[
    'cluster',
    'Majority',
    'filedEviction',
    'White Alone',
    'Black Alone',
    'Hispanic or Latino',
    'American Indian or Alaska Native Alone',
    'Asian Alone'
]].rename(columns={
    'cluster': 'cluster',
    'Majority': 'Majority',
    'filedEviction': 'filed_eviction',
    'White Alone': 'white',
    'Black Alone': 'black',
    'Hispanic or Latino': 'latino',
    'American Indian or Alaska Native Alone': 'native_american',
    'Asian Alone': 'asian'})

# Run ANOVA
Majority_anova = smf.ols(
    formula='filed_eviction ~ C(cluster) + white + black + latino + native_american + asian + C(Majority)+C(cluster):C(Majority)',
    data=ethical
).fit()

anova_table = sm.stats.anova_lm(Majority_anova, typ=2)
print(anova_table)

                               sum_sq      df         F    PR(>F)
C(cluster)                 281.890967     8.0  0.115364  0.998682
C(Majority)                 -0.116522     3.0 -0.000127  1.000000
C(cluster):C(Majority)     532.942782    24.0  0.072702  0.974605
white                        8.117486     1.0  0.026577  0.870522
black                      212.181047     1.0  0.694681  0.404709
latino                     346.295817     1.0  1.133773  0.287143
native_american            650.415368     1.0  2.129461  0.144700
asian                      415.930558     1.0  1.361757  0.243418
Residual                459987.567059  1506.0       NaN       NaN


/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 24, but rank is 3
  warnings.warn('covariance of constraints does not have full '


The anova result shows that the composition of ethical groups is not statistically significant to filed eviction cases between clusters.

# **majorHouseType**

In [22]:
renamed_df = fullDataset.rename(columns={
    'Single-family homes': 'single_family',
    'Small multi-unit buildings (2-4 units)': 'small_multi',
    'Larger apartment complexes (5+ units)': 'large_apartment',
    'Mobile homes, boats, RVs, etc.': 'mobile_other'
})


In [23]:

majorHouseType_anova = smf.ols(
    formula='filedEviction ~ C(cluster) + single_family + small_multi + large_apartment + mobile_other + C(majorHouseType)+C(majorHouseType):C(cluster)',
    data=renamed_df
).fit()

anova_table = sm.stats.anova_lm(majorHouseType_anova, typ=2)
print(anova_table)


                                     sum_sq      df             F    PR(>F)
C(cluster)                         0.000213     8.0  8.582327e-08  1.000000
C(majorHouseType)                 -0.000012     2.0 -1.971436e-08  1.000000
C(majorHouseType):C(cluster)    1144.704525    16.0  2.310147e-01  0.793756
single_family                    729.625818     1.0  2.355952e+00  0.125015
small_multi                       49.666888     1.0  1.603737e-01  0.688869
large_apartment                  569.159806     1.0  1.837810e+00  0.175410
mobile_other                       1.736254     1.0  5.606342e-03  0.940324
Residual                      468258.322936  1512.0           NaN       NaN


/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 7
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 2, but rank is 1
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 16, but rank is 2
  warnings.warn('covariance of constraints does not have full '


The interaction between `HouseType` and `cluster` is statistically significant in relation to filed eviction cases, as the p-value is below 0.05.

However, the individual ratios of house types are not statistically significant, indicating that the distribution of house types does not differ significantly between `clusters`.

**列表整理出所有的P_value**

In [24]:
import pandas as pd

## list of anova formula
anova_list = [
    sm.stats.anova_lm(median_anova, typ=2),
    sm.stats.anova_lm(eviction_anova, typ=2),
    sm.stats.anova_lm(burden_anova, typ=2),
    sm.stats.anova_lm(Majority_anova, typ=2),
    sm.stats.anova_lm(majorHouseType_anova, typ=2),
]

# merge the p-value
p_values_combined = pd.concat([anova['PR(>F)'] for anova in anova_list])

# dataframe
result_df = pd.DataFrame({
    'Variable': p_values_combined.index,
    'p_value': p_values_combined.values
})

# use 0.05 as a threshold for the significance
result_df['Significant'] = result_df['p_value'].apply(lambda p: '✅' if p < 0.05 else '❌')

result_df = result_df.drop_duplicates(subset='Variable').reset_index(drop=True)

display(result_df)



/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 7
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 9, but rank is 8
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 4
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have f

,Variable,p_value,Significant
0,C(cluster),0.000000e+00,✅
1,C(incomeLevel),1.666364e-02,✅
2,medianIncome,4.729738e-01,❌
3,medianIncome:C(incomeLevel),9.054343e-04,✅
4,Residual,NaN,❌
5,C(LiveType):C(cluster),1.389225e-05,✅
6,OwnerOccupied,5.499103e-02,❌
7,RenterOccupied,1.940977e-01,❌
8,C(burden),9.201499e-215,✅
9,C(cluster):C(burden),4.347135e-01,❌


# **抓出小於0.05的並建立模型**

In [25]:
# 選出顯著變數且排除 NaN 與 Residual
significant_vars = result_df[
    (result_df['p_value'] < 0.05) &
    (result_df['Variable'] != 'Residual') &
    (~result_df['Variable'].duplicated())
]['Variable'].tolist()

print("✅ 顯著變數：")
print(significant_vars)


✅ 顯著變數：
['C(cluster)', 'C(incomeLevel)', 'medianIncome:C(incomeLevel)', 'C(LiveType):C(cluster)', 'C(burden)', 'lowBurdan', 'moderateBurdan']


In [26]:
# 建立模型並跑 ANOVA
formula = 'filedEviction ~ ' + ' + '.join(significant_vars)
selected_model = smf.ols(formula=formula, data=fullDataset).fit()
selected_anova = sm.stats.anova_lm(selected_model, typ=2)

# 顯示結果
display(selected_anova)


/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 7
  warnings.warn('covariance of constraints does not have full '
/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 9, but rank is 8
  warnings.warn('covariance of constraints does not have full '


,sum_sq,df,F,PR(>F)
C(cluster),4.926163e+06,8.0,2046.781862,0.000000
C(incomeLevel),6.239928e+03,3.0,6.913709,0.000127
C(burden),7.481651e+02,3.0,0.828951,0.477894
C(LiveType):C(cluster),1.391870e+04,9.0,5.140544,0.000002
medianIncome:C(incomeLevel),4.039312e+03,4.0,3.356605,0.009579
lowBurdan,1.857175e+02,1.0,0.617313,0.432171
moderateBurdan,7.717470e+02,1.0,2.565239,0.109446
Residual,4.524755e+05,1504.0,NaN,NaN


## **Final model: 去掉小於0.05的再跑一次anova**

In [27]:
# 去掉 NaN 和保留 p < 0.05 的變數
final_vars = selected_anova[
    (selected_anova['PR(>F)'] < 0.05) &
    (selected_anova.index != 'Residual')
].index.tolist()
print("🎯 顯著變數（最終）:")
print(final_vars)


🎯 顯著變數（最終）:
['C(cluster)', 'C(incomeLevel)', 'C(LiveType):C(cluster)', 'medianIncome:C(incomeLevel)']


In [28]:
# Create the final formula (make sure to keep the C(...) format)
formula_final = 'filedEviction ~ C(cluster)+C(incomeLevel)+C(cluster):C(Majority)+medianIncome:C(incomeLevel)'
print("📌 Final formula:", formula_final)

# Build the model
final_model = smf.ols(formula=formula_final, data=fullDataset).fit()
final_anova = sm.stats.anova_lm(final_model, typ=2)

# Display the results
display(final_anova)


/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 8, but rank is 6
  warnings.warn('covariance of constraints does not have full '


📌 Final formula: filedEviction ~ C(cluster)+C(incomeLevel)+C(cluster):C(Majority)+medianIncome:C(incomeLevel)


/home/victortsai/miniconda3/envs/AZeviction/lib/python3.11/site-packages/statsmodels/base/model.py:1894: ValueWarning: covariance of constraints does not have full rank. The number of constraints is 27, but rank is 20
  warnings.warn('covariance of constraints does not have full '


,sum_sq,df,F,PR(>F)
C(cluster),1.858477e-03,8.0,7.711925e-07,1.000000
C(incomeLevel),4.761075e+03,3.0,5.268409e+00,0.001289
C(cluster):C(Majority),3.347224e+06,27.0,4.115444e+02,0.000000
medianIncome:C(incomeLevel),5.244431e+03,4.0,4.352453e+00,0.001674
Residual,4.530563e+05,1504.0,NaN,NaN


**Description:** In our final regression model, medianIncome and cluster are the top two most important factors in this model.

Thus, median income and majority house type in tracts between clusters are the statistically significant to filed eviction cases.

# **模型解釋力( final model最好)**

In [29]:
from sklearn.metrics import mean_squared_error
import numpy as np

def evaluate_model(model, name='model'):
    r2 = model.rsquared
    aic = model.aic
    y_true = model.model.endog
    y_pred = model.fittedvalues
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    return pd.Series({'R²': r2, 'AIC': aic, 'RMSE': rmse}, name=name)


In [30]:
# 比較幾個模型：最終精簡版、全版本等
results = pd.DataFrame([
    evaluate_model(final_model, 'Final Model'),
    evaluate_model(median_anova, 'Median Income Only'),
    evaluate_model(eviction_anova, 'LiveType Model'),
    evaluate_model(burden_anova, 'Burden Model'),
    evaluate_model(Majority_anova, 'Race & Majority'),
    evaluate_model(majorHouseType_anova, 'House Type Model'),
])

# 顯示
display(results.sort_values('AIC'))


,R²,AIC,RMSE
Final Model,0.920448,13129.380241,17.191153
Median Income Only,0.918261,13144.957683,17.425867
LiveType Model,0.918473,13144.969574,17.403215
Race & Majority,0.919231,13148.655939,17.322157
House Type Model,0.917778,13163.975003,17.477193
Burden Model,0.918066,13188.606875,17.446619


**Description:**
We use the R-square (ratio of variance is explained), AIC and RMSE (loss function) to evaluate the models.

Finally, we find the Final Model has the best performance on fitting filed eviction; the model can explain 91% of variance in training data, and least AIC indicated that the final model fitted well.